From 08587126fd29494bfa3dfb8ddb4630fffba890bb Mon Sep 17 00:00:00 2001
From: Dylan Yip <dylan.yip@xilinx.com>
Date: Tue, 9 Mar 2021 12:36:36 -0800
Subject: [PATCH 1765/1852] drm: xlnx: zynqmp: Only register DP CRTC if there
 is no other CRTC

commit 39650de1729d1b70d6ac03b78566c4da38453ba7 from
https://github.com/Xilinx/linux-xlnx.git

Currently the PS DP CRTC is automatically registered even if we are
using an external CRTC. So, only register the PS DP CRTC when there are
no other possible CRTCs/ports connected to PS DP. Same thing with
unregister.

Signed-off-by: Dylan Yip <dylan.yip@xilinx.com>
Signed-off-by: Jianqiang Chen <jianqiang.chen@xilinx.com>
Acked-by: Varunkumar Allagadapa <varunkumar.allagadapa@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/zynqmp_disp.c  |  7 +++++--
 drivers/gpu/drm/xlnx/zynqmp_dp.c    | 17 +++++++++++------
 drivers/gpu/drm/xlnx/zynqmp_dpsub.h |  1 +
 3 files changed, 17 insertions(+), 8 deletions(-)

diff --git a/drivers/gpu/drm/xlnx/zynqmp_disp.c b/drivers/gpu/drm/xlnx/zynqmp_disp.c
index 387883f04f09..b3ba23c34d8f 100644
--- a/drivers/gpu/drm/xlnx/zynqmp_disp.c
+++ b/drivers/gpu/drm/xlnx/zynqmp_disp.c
@@ -3101,14 +3101,17 @@ static int zynqmp_disp_create_crtc(struct zynqmp_disp *disp)
 	disp->xlnx_crtc.get_format = &zynqmp_disp_get_format;
 	disp->xlnx_crtc.get_align = &zynqmp_disp_get_align;
 	disp->xlnx_crtc.get_dma_mask = &zynqmp_disp_get_dma_mask;
-	xlnx_crtc_register(disp->drm, &disp->xlnx_crtc);
+	/* Only register the PS DP CRTC if there is no external port/CRTC */
+	if (!disp->dpsub->external_crtc_attached)
+		xlnx_crtc_register(disp->drm, &disp->xlnx_crtc);
 
 	return 0;
 }
 
 static void zynqmp_disp_destroy_crtc(struct zynqmp_disp *disp)
 {
-	xlnx_crtc_unregister(disp->drm, &disp->xlnx_crtc);
+	if (!disp->dpsub->external_crtc_attached)
+		xlnx_crtc_unregister(disp->drm, &disp->xlnx_crtc);
 	zynqmp_disp_crtc_destroy(&disp->xlnx_crtc.crtc);
 }
 
diff --git a/drivers/gpu/drm/xlnx/zynqmp_dp.c b/drivers/gpu/drm/xlnx/zynqmp_dp.c
index 48cec7aa2a6a..238cfbfad1e1 100644
--- a/drivers/gpu/drm/xlnx/zynqmp_dp.c
+++ b/drivers/gpu/drm/xlnx/zynqmp_dp.c
@@ -1681,20 +1681,16 @@ int zynqmp_dp_bind(struct device *dev, struct device *master, void *data)
 	struct drm_encoder *encoder = &dp->encoder;
 	struct drm_connector *connector = &dp->connector;
 	struct drm_device *drm = data;
-	struct device_node *port;
 	int ret;
 
 	if (!dp->num_lanes)
 		return 0;
 
 	encoder->possible_crtcs |= zynqmp_disp_get_crtc_mask(dpsub->disp);
-	for_each_child_of_node(dev->of_node, port) {
-		if (!port->name || of_node_cmp(port->name, "port"))
-			continue;
+	if (dpsub->external_crtc_attached)
 		encoder->possible_crtcs |=
 			drm_of_find_possible_crtcs(drm, dev->of_node);
-		break;
-	}
+
 	drm_encoder_init(drm, encoder, &zynqmp_dp_encoder_funcs,
 			 DRM_MODE_ENCODER_TMDS, NULL);
 	drm_encoder_helper_add(encoder, &zynqmp_dp_encoder_helper_funcs);
@@ -1819,6 +1815,7 @@ int zynqmp_dp_probe(struct platform_device *pdev)
 	struct zynqmp_dpsub *dpsub;
 	struct zynqmp_dp *dp;
 	struct resource *res;
+	struct device_node *port;
 	unsigned int i;
 	int irq, ret;
 
@@ -1907,6 +1904,14 @@ int zynqmp_dp_probe(struct platform_device *pdev)
 	dpsub->dp = dp;
 	dp->dpsub = dpsub;
 
+	for_each_child_of_node(pdev->dev.of_node, port) {
+		if (!port->name || of_node_cmp(port->name, "port"))
+			continue;
+
+		dpsub->external_crtc_attached = true;
+		break;
+	}
+
 	dev_dbg(dp->dev,
 		"ZynqMP DisplayPort Tx driver probed with %u phy lanes\n",
 		dp->num_lanes);
diff --git a/drivers/gpu/drm/xlnx/zynqmp_dpsub.h b/drivers/gpu/drm/xlnx/zynqmp_dpsub.h
index 6606beffee15..1adeff5b5811 100644
--- a/drivers/gpu/drm/xlnx/zynqmp_dpsub.h
+++ b/drivers/gpu/drm/xlnx/zynqmp_dpsub.h
@@ -22,6 +22,7 @@
 struct zynqmp_dpsub {
 	struct zynqmp_dp *dp;
 	struct zynqmp_disp *disp;
+	bool external_crtc_attached;
 	struct platform_device *master;
 };
 
-- 
2.31.1

